{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 1: Controls\n",
    "\n",
    "Write a Python script that proves that the lines of data in Germplasm.tsv, and LocusGene are in the same sequence, based on the AGI Locus Code (ATxGxxxxxx). (hint: This will help you decide how to load the data into the database)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The lines of data in Germplasm.tsv and LocusGene.tsv are in the same sequence \n"
     ]
    }
   ],
   "source": [
    "import csv\n",
    "\n",
    "def get_locus_entries(file):\n",
    "    Opened_file = open(file, 'r')\n",
    "    Data = csv.DictReader(Opened_file, delimiter = \"\\t\", quotechar='\"')\n",
    "    Locus=[]\n",
    "    for row in Data:\n",
    "        Locus.append(row[\"Locus\"])\n",
    "    Opened_file.close()\n",
    "    return(Locus)\n",
    "\n",
    "Locus_germplasm = get_locus_entries(\"Germplasm.tsv\")\n",
    "Locus_gene = get_locus_entries(\"LocusGene.tsv\")\n",
    "\n",
    "if Locus_germplasm == Locus_gene:\n",
    "    print(\"The lines of data in Germplasm.tsv and LocusGene.tsv are in the same sequence \")\n",
    "else:\n",
    "    print(\"The lines of data are in different sequences, or something went wrong\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, a function is defined to extract the Locus codes from any given file with a Locus column using the csv library.\n",
    "Then, Locus are extracted from both files.\n",
    "Finally, a simple if condition is set up to identify if all Locus codes in both files are equal, meaning that the lines of data are in the same sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Connected: root@mysql'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%load_ext sql\n",
    "#%config SqlMagic.autocommit=False #<- asi los cambios no son permanentes, por defecto se autocambia en la base de datos\n",
    "%sql mysql+pymysql://root:root@127.0.0.1:3306/mysql\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 290,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "8 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Database</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>information_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Prueba1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>Prueba2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>germplasm</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>mysql</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>performance_schema</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>sys</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>testing123</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('information_schema',),\n",
       " ('Prueba1',),\n",
       " ('Prueba2',),\n",
       " ('germplasm',),\n",
       " ('mysql',),\n",
       " ('performance_schema',),\n",
       " ('sys',),\n",
       " ('testing123',)]"
      ]
     },
     "execution_count": 290,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql show databases\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 2:  Design and create the database.  \n",
    "* It should have two tables - one for each of the two data files.\n",
    "* The two tables should be linked in a 1:1 relationship\n",
    "* you may use either sqlMagic or pymysql to build the database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 291,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The database 'Exam_2' was succesfully created\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import pymysql\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='mysql',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             autocommit=True,\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        # Read a single record\n",
    "        sql = \"CREATE database Exam_2\"\n",
    "        cursor.execute(sql)\n",
    "        print(\"The database 'Exam_2' was succesfully created\")\n",
    "        #connection.commit()\n",
    "        \n",
    "finally:\n",
    "    print(\"\")\n",
    "    #connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First the database 'Exam_2' is needed to store our 2 tables. The code above creates the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 292,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The tables 'Germplasm' and 'LocusGene' were succesfully created\n",
      "\n"
     ]
    }
   ],
   "source": [
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Exam_2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             autocommit=True,\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "try:\n",
    "    with connection.cursor() as cursor:\n",
    "        # Read a single record\n",
    "        sql = \"CREATE TABLE Germplasm(Locus VARCHAR(10) NOT NULL PRIMARY KEY, germplasm VARCHAR(40) NOT NULL, phenotype VARCHAR(500) NOT NULL, pubmed INTEGER NOT NULL)\"\n",
    "        cursor.execute(sql)\n",
    "        sql = \"CREATE TABLE LocusGene(Locus VARCHAR(10) NOT NULL PRIMARY KEY, Gene VARCHAR(10) NOT NULL, ProteinLength INTEGER NOT NULL)\"\n",
    "        cursor.execute(sql)\n",
    "        print(\"The tables 'Germplasm' and 'LocusGene' were succesfully created\")\n",
    "        #connection.commit()\n",
    "        \n",
    "finally:\n",
    "    print(\"\")\n",
    "    #connection.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The connection is changed to the new 'Exam_2' database, and with 2 querries, Germplasm and LocusGene tables are created in the database."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 293,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Locus</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed</th>\n",
       "        <th>Locus_1</th>\n",
       "        <th>Gene</th>\n",
       "        <th>ProteinLength</th>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 293,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql use Exam_2\n",
    "%sql SELECT * FROM Germplasm, LocusGene WHERE Germplasm.Locus = LocusGene.Locus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The tables are linked through the Locus column."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 3: Fill the database\n",
    "\n",
    "Using pymysql, create a Python script that reads the data from these files, and fills the database. There are a variety of strategies to accomplish this. I will give all strategies equal credit - do whichever one you are most confident with.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 294,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The table Germplasm was successfully filled\n",
      "\n",
      "The table LocusGene was successfully filled\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import csv, pymysql\n",
    "\n",
    "def Fill_table(file, table):\n",
    "    Opened_file = open(file, 'r')\n",
    "    Data = csv.DictReader(Opened_file, delimiter = \"\\t\", quotechar = '\"')\n",
    "    try:\n",
    "        with connection.cursor() as cursor:\n",
    "            if file == \"LocusGene.tsv\":\n",
    "                for row in Data:\n",
    "                    sql = \"INSERT INTO LocusGene (Locus, Gene, ProteinLength) VALUES ('{}', '{}', {})\".format(row[\"Locus\"], row[\"Gene\"], row[\"ProteinLength\"])\n",
    "                    cursor.execute(sql)\n",
    "            elif file == \"Germplasm.tsv\":\n",
    "                for row in Data:\n",
    "                    sql = \"INSERT INTO Germplasm (Locus, germplasm, phenotype, pubmed) VALUES ('{}', '{}', '{}', {})\".format(row[\"Locus\"], row[\"germplasm\"], row[\"phenotype\"], int(row[\"pubmed\"]))\n",
    "                    cursor.execute(sql)\n",
    "            print(\"The table {} was successfully filled\".format(table))\n",
    "    finally:\n",
    "        print(\"\")\n",
    "    Opened_file.close()\n",
    "\n",
    "# =========== Connexion:\n",
    "connection = pymysql.connect(host='localhost',\n",
    "                             user='root',\n",
    "                             password='root',\n",
    "                             db='Exam_2',\n",
    "                             charset='utf8mb4',  # note utf8... this is important for unusual characters!\n",
    "                             autocommit=True,\n",
    "                             cursorclass=pymysql.cursors.DictCursor)\n",
    "\n",
    "# ========== Germplasm table:\n",
    "Fill_table(\"Germplasm.tsv\", \"Germplasm\")\n",
    "\n",
    "# =========== LocusGene:\n",
    "Fill_table(\"LocusGene.tsv\", \"LocusGene\")\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this problem, first a function is defined to fill a given table from a given file, and this function is used to fill both tables."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "0 rows affected.\n",
      " * mysql+pymysql://root:***@127.0.0.1:3306/mysql\n",
      "32 rows affected.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table>\n",
       "    <tr>\n",
       "        <th>Locus</th>\n",
       "        <th>germplasm</th>\n",
       "        <th>phenotype</th>\n",
       "        <th>pubmed</th>\n",
       "        <th>Locus_1</th>\n",
       "        <th>Gene</th>\n",
       "        <th>ProteinLength</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>CS3828</td>\n",
       "        <td>Increased abundance of miRNA precursors.</td>\n",
       "        <td>17369351</td>\n",
       "        <td>AT1G01040</td>\n",
       "        <td>DCL1</td>\n",
       "        <td>332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>lhy-101</td>\n",
       "        <td>The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.</td>\n",
       "        <td>16891401</td>\n",
       "        <td>AT1G01060</td>\n",
       "        <td>LHY</td>\n",
       "        <td>290</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>SALK_058629</td>\n",
       "        <td>hypersensitive to low potassium media</td>\n",
       "        <td>17486125</td>\n",
       "        <td>AT1G01140</td>\n",
       "        <td>CIPK9</td>\n",
       "        <td>223</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>SALK_012400C</td>\n",
       "        <td>fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.</td>\n",
       "        <td>18199744</td>\n",
       "        <td>AT1G01220</td>\n",
       "        <td>FKGP</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>SALK_042433</td>\n",
       "        <td>Multiple straight hairs</td>\n",
       "        <td>16367956</td>\n",
       "        <td>AT2G03720</td>\n",
       "        <td>MRH6</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>gek1-1</td>\n",
       "        <td>Ethanol hypersensitivity.</td>\n",
       "        <td>15215505</td>\n",
       "        <td>AT2G03800</td>\n",
       "        <td>GEK1</td>\n",
       "        <td>196</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>xerico</td>\n",
       "        <td>Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.</td>\n",
       "        <td>17933900</td>\n",
       "        <td>AT2G04240</td>\n",
       "        <td>XERICO</td>\n",
       "        <td>256</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>pot1-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17627276</td>\n",
       "        <td>AT2G05210</td>\n",
       "        <td>POT1A</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>rpk2-2</td>\n",
       "        <td>The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.</td>\n",
       "        <td>17419837</td>\n",
       "        <td>AT3G02130</td>\n",
       "        <td>RPK2</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>afp4-1</td>\n",
       "        <td>Decreased germination on high concentrations of glucose and sorbitol.</td>\n",
       "        <td>18484180</td>\n",
       "        <td>AT3G02140</td>\n",
       "        <td>TMAC2</td>\n",
       "        <td>300</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>rgp1-1</td>\n",
       "        <td>rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.</td>\n",
       "        <td>21478444</td>\n",
       "        <td>AT3G02230</td>\n",
       "        <td>RGP1</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>tir3-1 RGLG1:rglg1 rglg2</td>\n",
       "        <td>The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.</td>\n",
       "        <td>17586653</td>\n",
       "        <td>AT3G02260</td>\n",
       "        <td>BIG</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>sep2-1</td>\n",
       "        <td>Non-described subtle phenotype.</td>\n",
       "        <td>10821278</td>\n",
       "        <td>AT3G02310</td>\n",
       "        <td>SEP2</td>\n",
       "        <td>175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>atnbs1-1</td>\n",
       "        <td>Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.</td>\n",
       "        <td>17672843</td>\n",
       "        <td>AT3G02680</td>\n",
       "        <td>NBS1</td>\n",
       "        <td>190</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>CS3816</td>\n",
       "        <td>The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.</td>\n",
       "        <td>17568770</td>\n",
       "        <td>AT3G02850</td>\n",
       "        <td>SKOR</td>\n",
       "        <td>234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>vtc4-1</td>\n",
       "        <td>ascorbate deficient</td>\n",
       "        <td>16595667</td>\n",
       "        <td>AT3G02870</td>\n",
       "        <td>VTC4</td>\n",
       "        <td>311</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>hdg8-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16778018</td>\n",
       "        <td>AT3G03260</td>\n",
       "        <td>HDG8</td>\n",
       "        <td>194</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>pdd17</td>\n",
       "        <td>Defective pollen development.</td>\n",
       "        <td>19237690</td>\n",
       "        <td>AT4G14790</td>\n",
       "        <td>SUV3</td>\n",
       "        <td>312</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>bmy1-2</td>\n",
       "        <td>Plants cold-shocked for 6h have an increased starch content compared to wildtype.</td>\n",
       "        <td>16297066</td>\n",
       "        <td>AT4G15210</td>\n",
       "        <td>BAM5</td>\n",
       "        <td>313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>cla1-1</td>\n",
       "        <td>Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.</td>\n",
       "        <td>10982425</td>\n",
       "        <td>AT4G15560</td>\n",
       "        <td>DXS</td>\n",
       "        <td>219</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>maa3</td>\n",
       "        <td>Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.</td>\n",
       "        <td>18772186</td>\n",
       "        <td>AT4G15570</td>\n",
       "        <td>MAA3</td>\n",
       "        <td>294</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>Athspb-2</td>\n",
       "        <td>Early flowering, reduced fertility, aborted seeds.</td>\n",
       "        <td>20388662</td>\n",
       "        <td>AT4G15802</td>\n",
       "        <td>HSBP</td>\n",
       "        <td>254</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>esd4-2</td>\n",
       "        <td>Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.</td>\n",
       "        <td>17513499</td>\n",
       "        <td>AT4G15880</td>\n",
       "        <td>ESD4</td>\n",
       "        <td>265</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>prz1-1</td>\n",
       "        <td>Altered response to auxin and cytokinin</td>\n",
       "        <td>12747832</td>\n",
       "        <td>AT4G16420</td>\n",
       "        <td>ADA2B</td>\n",
       "        <td>279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>atint4-2</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>16603666</td>\n",
       "        <td>AT4G16480</td>\n",
       "        <td>INT4</td>\n",
       "        <td>284</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>pas2-3</td>\n",
       "        <td>Segregates 25% embryo lethal.</td>\n",
       "        <td>18799749</td>\n",
       "        <td>AT5G10480</td>\n",
       "        <td>PAS2</td>\n",
       "        <td>301</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>plt3-1</td>\n",
       "        <td>Short roots and shortened root meristem.</td>\n",
       "        <td>17960244</td>\n",
       "        <td>AT5G10510</td>\n",
       "        <td>AIL6</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>kns2</td>\n",
       "        <td>Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.</td>\n",
       "        <td>18779216</td>\n",
       "        <td>AT5G11110</td>\n",
       "        <td>SPS2</td>\n",
       "        <td>232</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>hy5-101</td>\n",
       "        <td>Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.</td>\n",
       "        <td>16891401</td>\n",
       "        <td>AT5G11260</td>\n",
       "        <td>HY5</td>\n",
       "        <td>221</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>myb3r4-1</td>\n",
       "        <td>No visible phenotype.</td>\n",
       "        <td>17287251</td>\n",
       "        <td>AT5G11510</td>\n",
       "        <td>MYB3R-4</td>\n",
       "        <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>pyd2-2</td>\n",
       "        <td>The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.</td>\n",
       "        <td>19413687</td>\n",
       "        <td>AT5G12200</td>\n",
       "        <td>PYD2</td>\n",
       "        <td>310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>crn-1</td>\n",
       "        <td>Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.</td>\n",
       "        <td>12345678</td>\n",
       "        <td>AT5G13290</td>\n",
       "        <td>CRN</td>\n",
       "        <td>189</td>\n",
       "    </tr>\n",
       "</table>"
      ],
      "text/plain": [
       "[('AT1G01040', 'CS3828', 'Increased abundance of miRNA precursors.', 17369351, 'AT1G01040', 'DCL1', 332),\n",
       " ('AT1G01060', 'lhy-101', 'The mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.', 16891401, 'AT1G01060', 'LHY', 290),\n",
       " ('AT1G01140', 'SALK_058629', 'hypersensitive to low potassium media', 17486125, 'AT1G01140', 'CIPK9', 223),\n",
       " ('AT1G01220', 'SALK_012400C', 'fkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ s ... (41 characters truncated) ... henotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.', 18199744, 'AT1G01220', 'FKGP', 190),\n",
       " ('AT2G03720', 'SALK_042433', 'Multiple straight hairs', 16367956, 'AT2G03720', 'MRH6', 189),\n",
       " ('AT2G03800', 'gek1-1', 'Ethanol hypersensitivity.', 15215505, 'AT2G03800', 'GEK1', 196),\n",
       " ('AT2G04240', 'xerico', 'Resistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.', 17933900, 'AT2G04240', 'XERICO', 256),\n",
       " ('AT2G05210', 'pot1-1', 'No visible phenotype.', 17627276, 'AT2G05210', 'POT1A', 221),\n",
       " ('AT3G02130', 'rpk2-2', 'The homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.', 17419837, 'AT3G02130', 'RPK2', 284),\n",
       " ('AT3G02140', 'afp4-1', 'Decreased germination on high concentrations of glucose and sorbitol.', 18484180, 'AT3G02140', 'TMAC2', 300),\n",
       " ('AT3G02230', 'rgp1-1', 'rgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.', 21478444, 'AT3G02230', 'RGP1', 301),\n",
       " ('AT3G02260', 'tir3-1 RGLG1:rglg1 rglg2', 'The triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.', 17586653, 'AT3G02260', 'BIG', 279),\n",
       " ('AT3G02310', 'sep2-1', 'Non-described subtle phenotype.', 10821278, 'AT3G02310', 'SEP2', 175),\n",
       " ('AT3G02680', 'atnbs1-1', 'Significantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.', 17672843, 'AT3G02680', 'NBS1', 190),\n",
       " ('AT3G02850', 'CS3816', 'The skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.', 17568770, 'AT3G02850', 'SKOR', 234),\n",
       " ('AT3G02870', 'vtc4-1', 'ascorbate deficient', 16595667, 'AT3G02870', 'VTC4', 311),\n",
       " ('AT3G03260', 'hdg8-1', 'No visible phenotype.', 16778018, 'AT3G03260', 'HDG8', 194),\n",
       " ('AT4G14790', 'pdd17', 'Defective pollen development.', 19237690, 'AT4G14790', 'SUV3', 312),\n",
       " ('AT4G15210', 'bmy1-2', 'Plants cold-shocked for 6h have an increased starch content compared to wildtype.', 16297066, 'AT4G15210', 'BAM5', 313),\n",
       " ('AT4G15560', 'cla1-1', 'Mutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.', 10982425, 'AT4G15560', 'DXS', 219),\n",
       " ('AT4G15570', 'maa3', 'Homozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.', 18772186, 'AT4G15570', 'MAA3', 294),\n",
       " ('AT4G15802', 'Athspb-2', 'Early flowering, reduced fertility, aborted seeds.', 20388662, 'AT4G15802', 'HSBP', 254),\n",
       " ('AT4G15880', 'esd4-2', 'Decreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.', 17513499, 'AT4G15880', 'ESD4', 265),\n",
       " ('AT4G16420', 'prz1-1', 'Altered response to auxin and cytokinin', 12747832, 'AT4G16420', 'ADA2B', 279),\n",
       " ('AT4G16480', 'atint4-2', 'No visible phenotype.', 16603666, 'AT4G16480', 'INT4', 284),\n",
       " ('AT5G10480', 'pas2-3', 'Segregates 25% embryo lethal.', 18799749, 'AT5G10480', 'PAS2', 301),\n",
       " ('AT5G10510', 'plt3-1', 'Short roots and shortened root meristem.', 17960244, 'AT5G10510', 'AIL6', 310),\n",
       " ('AT5G11110', 'kns2', 'Defects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.', 18779216, 'AT5G11110', 'SPS2', 232),\n",
       " ('AT5G11260', 'hy5-101', 'Under FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.', 16891401, 'AT5G11260', 'HY5', 221),\n",
       " ('AT5G11510', 'myb3r4-1', 'No visible phenotype.', 17287251, 'AT5G11510', 'MYB3R-4', 336),\n",
       " ('AT5G12200', 'pyd2-2', 'The pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.', 19413687, 'AT5G12200', 'PYD2', 310),\n",
       " ('AT5G13290', 'crn-1', 'Increased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.', 12345678, 'AT5G13290', 'CRN', 189)]"
      ]
     },
     "execution_count": 295,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "%sql use Exam_2\n",
    "%sql SELECT * FROM Germplasm, LocusGene WHERE Germplasm.Locus = LocusGene.Locus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Problem 4: Create reports, written to a file\n",
    "\n",
    "1. Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "\n",
    "2. Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "3. Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "4. Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "When creating reports 2 and 3, remember the \"Don't Repeat Yourself\" rule! \n",
    "\n",
    "All reports should be written to **the same file**.  You may name the file anything you wish."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 296,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Report nº 1 was saved in the file Full_report.tsv\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import csv, io, pymysql\n",
    "\n",
    "# ============== Functions used in all reports:\n",
    "\n",
    "def Check_output():\n",
    "    '''Function that prints the content of the file \"Full_report.tsv\" when called'''\n",
    "    File = io.open(\"Full_report.tsv\")\n",
    "    print(File.read())\n",
    "    File.close()\n",
    "\n",
    "def Write_report(mode, table, report_number):\n",
    "    '''The two modes are \"w\" to write and \"a\" to append\n",
    "    This function accepts a list of dictionaries to write \n",
    "    the different reports in the file: Full_report.tsv'''\n",
    "    \n",
    "    Full_report = open('Full_report.tsv', mode, newline='')\n",
    "    if mode == \"w\": Full_report.write('#Exam Report \\n#Report nº1: \\n')\n",
    "    else: Full_report.write('\\n#Report nº{}: \\n'.format(report_number))\n",
    "    Report_writer = csv.DictWriter(Full_report, delimiter=\"\\t\", quotechar='\"', fieldnames=table[0].keys())\n",
    "    Report_writer.writeheader()    \n",
    "    Report_writer.writerows(table)\n",
    "    Full_report.close()\n",
    "    print(\"Report nº {} was saved in the file Full_report.tsv\\n\".format(report_number))\n",
    "  \n",
    "\n",
    "# === Report 1: Create a report that shows the full, joined, content of the two database tables (including a header line)\n",
    "def generic_querry(string):\n",
    "    connection = pymysql.connect(host='localhost',\n",
    "                                 user='root',\n",
    "                                 password='root',\n",
    "                                 db='Exam_2',\n",
    "                                 charset='utf8mb4', \n",
    "                                 autocommit=True,\n",
    "                                 cursorclass=pymysql.cursors.DictCursor)\n",
    "    try:\n",
    "        with connection.cursor() as cursor:\n",
    "            sql = string\n",
    "            cursor.execute(sql)\n",
    "            Linked_databases = cursor.fetchall()\n",
    "            # The second 'Locus' column (LocusGene.Locus) is removed, for clarity:\n",
    "            for row in Linked_databases:\n",
    "                row.pop(\"LocusGene.Locus\")\n",
    "            return(Linked_databases)\n",
    "    finally:\n",
    "        print(\"\")\n",
    "    \n",
    "Linked_databases = generic_querry(\"\"\"SELECT * FROM Germplasm, LocusGene WHERE Germplasm.Locus = LocusGene.Locus\"\"\")\n",
    "Write_report(mode='w', table=Linked_databases, report_number=1)\n",
    "\n",
    "# First output check:\n",
    "#Check_output()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For all reports, Check_output is a function that prints the current state of the full_report file.\n",
    "Write report is a function that takes a list of dictionaries(1 or more) and appends the report to the full report file.\n",
    "\n",
    "In the first report, we select everything and link both tables through the column Locus. The linked tables are stored in variable Linked_databases. We remove LocusGene.Gene from the linked tables because the column is repeated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 283,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Report nº 2 was saved in the file Full_report.tsv\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# ========= Report 2: Create a joined report that only includes the Genes SKOR and MAA3\n",
    "\n",
    "Linked_databases = generic_querry(\"\"\"SELECT * FROM Germplasm, LocusGene WHERE Germplasm.Locus = LocusGene.Locus \n",
    "        AND (LocusGene.Gene = 'SKOR' OR LocusGene.Gene = 'MAA3')\"\"\")\n",
    "Write_report(\"a\", Linked_databases, 2)\n",
    "\n",
    "# Second output check\n",
    "#Check_output()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For the second report, we add to the querry a condition so that both genes are selected with an OR between brackets and we write the report"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Report nº 3 was saved in the file Full_report.tsv\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# ========= Report 3: Create a report that counts the number of entries for each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "def Extract_data_each_chromosome(function, header_text, table):\n",
    "    connection = pymysql.connect(host='localhost',\n",
    "                                 user='root',\n",
    "                                 password='root',\n",
    "                                 db='Exam_2',\n",
    "                                 charset='utf8mb4', \n",
    "                                 autocommit=True,\n",
    "                                 cursorclass=pymysql.cursors.DictCursor)\n",
    "    try:\n",
    "        with connection.cursor() as cursor:\n",
    "            Chromosomes = []\n",
    "            for chromosome in [1,2,3,4,5]:\n",
    "                sql = \"\"\"SELECT {} AS '{} {}' FROM {} WHERE Locus REGEXP 'AT{}'\"\"\".format(function, header_text, chromosome, table, chromosome)\n",
    "                cursor.execute(sql)\n",
    "                Chromosomes.append(cursor.fetchall()[0])\n",
    "    finally:\n",
    "        print(\"\")\n",
    "    # We change the output so that instead of having a list of 5 dictionaries, \n",
    "    # it is a list containing a dictionary with 5 keys and values.\n",
    "    Results={}\n",
    "    for chromosome in Chromosomes:\n",
    "        Results.update(chromosome)\n",
    "    return([Results])\n",
    "\n",
    "# ==== Main code:\n",
    "\n",
    "Count = Extract_data_each_chromosome(\"COUNT(*)\", \"Counts in chromosome\", \"Germplasm\")\n",
    "Write_report(\"a\", Count, 3)\n",
    "\n",
    "# Third output check:\n",
    "#Check_output()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For both reports 3 and 4, we define the function Extract_data_each_chromosome that can accept a function to be applied in the querry (count or avg). This function can apply the given function to results from the 5 different chromosomes thanks to the REGEXP 'AT{}'. This way, we can observe the chromosomes independently. the function outputs a list containing a single dictionary with all information."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 285,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Report nº 4 was saved in the file Full_report.tsv\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# ========= Report 4: Create a report that shows the average protein length for the genes on each Chromosome (AT1Gxxxxxx to AT5Gxxxxxxx)\n",
    "\n",
    "Average=Extract_data_each_chromosome(\"AVG(LocusGene.ProteinLength)\", \"Average protein length in chromosome\", \"LocusGene\")\n",
    "Write_report(\"a\", Average, 4)\n",
    "\n",
    "# Fourth output check:\n",
    "#Check_output()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 286,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "#Exam Report \n",
      "#Report nº1: \n",
      "Locus\tgermplasm\tphenotype\tpubmed\tGene\tProteinLength\n",
      "AT1G01040\tCS3828\tIncreased abundance of miRNA precursors.\t17369351\tDCL1\t332\n",
      "AT1G01060\tlhy-101\tThe mutant plants are hypersensitive to both FRc and Rc light treatments in hypocotyl elongation and exhibits a small reciprocal enlargement in cotyledon area, albeit not statistically significant.\t16891401\tLHY\t290\n",
      "AT1G01140\tSALK_058629\thypersensitive to low potassium media\t17486125\tCIPK9\t223\n",
      "AT1G01220\tSALK_012400C\tfkgp-1 mutants have about 40 times more L-fucose than wild type Arabidopsis plants, but the levels of other monosaccharides do not appear to differ significantly in the mutants. No obvious phenotypic abnormalities were observed in the fkgp-1 mutants, nor were any differences in the sugar composition of cell wall polysaccharides detected.\t18199744\tFKGP\t190\n",
      "AT2G03720\tSALK_042433\tMultiple straight hairs\t16367956\tMRH6\t189\n",
      "AT2G03800\tgek1-1\tEthanol hypersensitivity.\t15215505\tGEK1\t196\n",
      "AT2G04240\txerico\tResistant to exogenous ABA. Seeds contained lower amounts of endogenous ABA than wildtype.\t17933900\tXERICO\t256\n",
      "AT2G05210\tpot1-1\tNo visible phenotype.\t17627276\tPOT1A\t221\n",
      "AT3G02130\trpk2-2\tThe homozygous progeny is indistinguishable from wild-type plants during vegetative growth but showed several morphological alterations after bolting. These plants displayed enhanced inflorescence branching and formed three times as many siliques and flowers as did wild-type plants.\t17419837\tRPK2\t284\n",
      "AT3G02140\tafp4-1\tDecreased germination on high concentrations of glucose and sorbitol.\t18484180\tTMAC2\t300\n",
      "AT3G02230\trgp1-1\trgp1-1 mutants have significantly lower levels of UDP-L-arabinose mutase activity compared to wild-type plants and significantly lower levels of arabinose in their cell walls.\t21478444\tRGP1\t301\n",
      "AT3G02260\ttir3-1 RGLG1:rglg1 rglg2\tThe triple homozygous progeny has low viability, accumulated anthocyanin, and all plants died before shoot emergence.\t17586653\tBIG\t279\n",
      "AT3G02310\tsep2-1\tNon-described subtle phenotype.\t10821278\tSEP2\t175\n",
      "AT3G02680\tatnbs1-1\tSignificantly smaller when grown in the presence of methyl methanosulfonate (MMS) with root growth.  Normal growth under standard growth conditions.\t17672843\tNBS1\t190\n",
      "AT3G02850\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\tSKOR\t234\n",
      "AT3G02870\tvtc4-1\tascorbate deficient\t16595667\tVTC4\t311\n",
      "AT3G03260\thdg8-1\tNo visible phenotype.\t16778018\tHDG8\t194\n",
      "AT4G14790\tpdd17\tDefective pollen development.\t19237690\tSUV3\t312\n",
      "AT4G15210\tbmy1-2\tPlants cold-shocked for 6h have an increased starch content compared to wildtype.\t16297066\tBAM5\t313\n",
      "AT4G15560\tcla1-1\tMutant seeds grown on medium supplemented with non-phosphorylated synthetic 1-deoxy-D-xylulose (DX) develop green leaves.\t10982425\tDXS\t219\n",
      "AT4G15570\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\tMAA3\t294\n",
      "AT4G15802\tAthspb-2\tEarly flowering, reduced fertility, aborted seeds.\t20388662\tHSBP\t254\n",
      "AT4G15880\tesd4-2\tDecreased mRNA levels of the floral repressors FLC and MAF4 and increased mRNA levels of the floral activators FT and SOC1.\t17513499\tESD4\t265\n",
      "AT4G16420\tprz1-1\tAltered response to auxin and cytokinin\t12747832\tADA2B\t279\n",
      "AT4G16480\tatint4-2\tNo visible phenotype.\t16603666\tINT4\t284\n",
      "AT5G10480\tpas2-3\tSegregates 25% embryo lethal.\t18799749\tPAS2\t301\n",
      "AT5G10510\tplt3-1\tShort roots and shortened root meristem.\t17960244\tAIL6\t310\n",
      "AT5G11110\tkns2\tDefects are specific to pollen exine structure. Smaller mesh size in the exine structure. Increased number of baculae. Fully fertile.\t18779216\tSPS2\t232\n",
      "AT5G11260\thy5-101\tUnder FRc conditions, the length mutant hypocotyls is increased compared to that of wild-type plants.  Under Rc conditions, the hypocotyl length is also increased and the cotyledon area is smaller.\t16891401\tHY5\t221\n",
      "AT5G11510\tmyb3r4-1\tNo visible phenotype.\t17287251\tMYB3R-4\t336\n",
      "AT5G12200\tpyd2-2\tThe pyd2-2 mutant has a wild-type appearance under normal growth conditions. Pyrimidine nucleotide and uridine levels are not changed in the mutant, but uracil levels are increased. These mutants cannot grow normally when uracil is provided as a sole nitrogen source.\t19413687\tPYD2\t310\n",
      "AT5G13290\tcrn-1\tIncreased meristem size. Vegetative meristems are are 30% larger than wild type. After bolting inflorescence meristems are enlarged and occasionally fasciated. Flowers occasionally produce extra organs in the first 3.\t12345678\tCRN\t189\n",
      "\n",
      "#Report nº2: \n",
      "Locus\tgermplasm\tphenotype\tpubmed\tGene\tProteinLength\n",
      "AT3G02850\tCS3816\tThe skor-1 mutant is sensitive to toxic cations in addition to K+ depletion.\t17568770\tSKOR\t234\n",
      "AT4G15570\tmaa3\tHomozygotes are not recovered. Female gametophyte development is delayed and asynchronous. During fertilization, fusion of polar nuclei does not occur. Polar nuclei nucloeli are smaller than WT.\t18772186\tMAA3\t294\n",
      "\n",
      "#Report nº3: \n",
      "Counts in chromosome 1\tCounts in chromosome 2\tCounts in chromosome 3\tCounts in chromosome 4\tCounts in chromosome 5\n",
      "4\t4\t9\t8\t7\n",
      "\n",
      "#Report nº4: \n",
      "Average protein length in chromosome 1\tAverage protein length in chromosome 2\tAverage protein length in chromosome 3\tAverage protein length in chromosome 4\tAverage protein length in chromosome 5\n",
      "258.7500\t215.5000\t252.0000\t277.5000\t271.2857\n",
      "\n"
     ]
    }
   ],
   "source": [
    "Check_output()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
